packages/zoe/src/zoe.js

```diff
@@ -440,84 +440,94 @@ const makeZoe = (additionalEndowments = {}) => {
      * @type {ContractFacet}
      */
     const contractFacet = harden({
-      reallocate: (offerHandles, newAllocations, sparseKeywords) => {
+      // reallocate takes two parallel arrays: offerHandles and newAllocations.
+      // In addition, sparseKeywords may be provided or may be empty.
+      // Each of the rows in newAllocations gives a collection of keyword-amount
+      // pairs where the amount should replace the old amount for that keyword.
+      // The reallocation will only succeed if the amounts specified have the
+      // same total value as the current total amount for those keywords on
+      // those offers, and 'offer safety' continues to be satisfied. The amounts
+      // for those brands on other keywords don't need to be compared.
+      // If sparseKeywords is a single array of keywords, only
+      // amounts for those keywords will be updated across all the
+      // newAllocations, and any other keyword will retain its current
+      // allocation. If sparseKeywords is an Array of Arrays, then it must be
+      // the same length as offerHandles, and each row will be applied
+      // separately to the respective current and newAllocations. If
+      // sparseKeywords is undefined, then newAllocations will represent a
+      // complete set of replacement values for the current allocations.
+      reallocate: (offerHandles, newAllocations, sparseKeywordsIn) => {
```

Comment:
I think the distinction between undefined and a nested array is needed. The callers should be able to apply

 * a single list of keywords across all offers
 * no filtering, so all keywords are checked
 * a different list of keywords for each offer.

If we leave out the third case, then a contract with many keywords that allows the keywords to vary across users (Dean seems to think this is a plausible model for multipool autoswap) wouldn't be able to reallocate for a few columns. 

I could see splitting it into 

 * `reallocate(offerHandles, newAllocations)`
 * `reallocateSparse(offerHandles, newAllocations, sparseKeywordsMatrix)`

Where the second case would allow either a 1- or 2-level array, or further dividing that case into `reallocateSparseArray()` and `reallocateSparseMatrix()`.

Do any of these seem more palatable than the current interface?

Reply:
> If we leave out the third case, then a contract with many keywords that allows the keywords to vary across users (Dean seems to think this is a plausible model for multipool autoswap) wouldn't be able to reallocate for a few columns.

I don't think this is right. Offer safety is comparing a users' wants and gives to a users' allocation, so it doesn't matter whether the keywords vary per offer (we're never comparing offers to each other). Rights conservation is adding up amounts based on brand, so keywords shouldn't matter there either. So actually, reallocate should not have any sense of objective keywords, meaning that we can simply have two parameters: `offerHandles` and `newAllocations` (the first of your split). I think the second case is unnecessary.

Reply:
There are two parts of what reallocation does. The first is the actual reallocation. I think we need all three cases for the reallocation. Reallocation requests come in three flavors:
 * No filtering: the parameters describe the entire state of the new allocation for each offer
 * a single list of keywords: the reallocation applies to only those keywords across all the offers
 * a different list of keywords for each offer: the reallocation applies keywords separately to each offer

Checking rights conservation doesn't care how the give and want are divided among keywords. 

I think offer safety does. If an offer says it  wants { a: 37, b: 20 }, then, even if the issuers are the same, a reallocation to { a: 58 } doesn't match the offer. Is this the wrong interpretation of what we mean by offer safety?